lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,234 @@
+        except FileNotFoundError:
+            raise OpenSearchPluginError(f"Plugin is missing - not found: {plugin_props_path}")
+        except PermissionError:
+            raise OpenSearchPluginError(
+                f"Plugin incorrectly installed - permission denied: {plugin_props_path}"
+            )
+        except Exception:
+            raise OpenSearchPluginError("Plugin Error - Unknown reason")
+        return self._properties._properties["version"]
+
+    @property
+    def dependencies(self) -> List[str]:
+        """Returns a list of plugin name dependencies."""
+        return self._depends_on
+
+    @abstractmethod
+    def install(self) -> Any:
+        """Executes any additional steps needed for the installation.
+
+        Should only execute steps that are not covered in PluginManager.
+        """
+        pass
+
+    @property
+    def is_installed(self) -> bool:
+        """Check if plugin is installed by checking if plugin properties file exists."""
+        try:
+            # Consider it is installed if properties file is present and readable.
+            _ = self.version
+        except Exception:
+            return False
+        return True
+
+    @abstractmethod
+    def config(self) -> Dict[str, Dict[str, str]]:
+        """Returns a dict containing all the configuration needed to be applied in the form.
+
+        Format:
+        {
+            "opensearch.yml": {...},
+            "keystore": {...},
+        }
+        """
+        pass
+
+    @property
+    def is_enabled(self) -> bool:
+        """Runs any specific steps to check if the plugin is enabled."""
+        return self._enabled
+
+    def _is_enabled(self, opensearch_config: Dict[str, Any]) -> None:
+        """Checks if the plugin is enabled based on configs set.
+
+        This method is ran only once, at initialization and should set self._enabled.
+        """
+        self._enabled = False
+
+    def needs_upgrade(self) -> bool:
+        """Returns if the plugin needs an upgrade or not.
+
+        Needs upgrade must be set if an upgrade on the charm happens and plugins must
+        be updated.
+
+        Consider overriding this method only if the plugin needs an special care at upgrade.
+        """
+        current_version = self.version
+        num_points = len(self._os_version.split("."))
+        return self._os_version != current_version[:num_points]
+
+    @abstractmethod
+    def upgrade(self, uri: str) -> Any:
+        """Runs specific extra steps for upgrade for this plugin."""
+        pass
+
+    def disable(self) -> Dict[str, Any]:
+        """Returns a dict containing different config changes.
+
+        The dict should return:
+        (1) all the configs to remove
+        (2) all the configs to add
+        (3) all the keystore values to be remmoved.
+        """
+        return {"to_remove_opensearch": [], "to_add_opensearch": [], "to_remove_keystore": []}
+
+    @property
+    def status(self) -> PluginState:
```

Comment:
We cannot track the status here, as this class is just a spec of who the plugin actually is.

This method should move to plugin manager.